### PR TITLE
POL 395 azure mca reservation recommendations/expiration

### PR DESCRIPTION
**Pull request by gopinathmahakud1**
### Description

Azure mca reservation recommendations/expiration new policy

### Issues Resolved

- Creating a new policy for azure MCA  reservation recommendations and expiration

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/60083e68b6b47d0001c87194
image
https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/60083d7c67016a000156f991
image

### CODE REVIEW

**Review by cdwilhelm — CHANGES_REQUESTED** [review]
looks very good.  just a few minor changes.

File: cost/azure/reserved_instances/mca_expiration/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# Azure Expiring Reserved Instances
+
+## What it does
+
+This Policy Template leverages the [Azure EA API](https://docs.microsoft.com/en-us/rest/api/reserved-vm-instances/reservationorder/list). It will notify only if expiration is within the time frame specified in `Number of days to prior to expiration date to trigger incident` field. It will email the user specified in Email addresses of the recipients you wish to notify.
```
[the remark this answers is not in the document]
Comment (gopinathmahakud1) [reply cost/azure/reserved_instances/mca_expiration/README.md]: I have addressed.

File: cost/azure/reserved_instances/mca_expiration/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+This policy uses [credentials](https://docs.rightscale.com/policies/users/guides/credential_management.html)
+for connecting to the cloud -- in order to apply this policy you must have a credential registered in the system that is compatible with this policy. If there are no
+credentials listed when you apply the policy, please contact your cloud admin and ask them to register a credential that is compatible with this policy. The information below should be consulted when creating the credential.
+
+### Credential configuration
+
+For administrators [creating and managing credentials](https://docs.rightscale.com/policies/users/guides/credential_management.html) to use with this policy, the following information is needed: 
+
+Provider tag value to match this policy: `azure`
```
[the remark this answers is not in the document]
Comment (gopinathmahakud1) [reply cost/azure/reserved_instances/mca_expiration/README.md]: I have addressed.

File: cost/azure/reserved_instances/mca_expiration/azure_reserved_instance_expiration.pt [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      field "quantity" do
+        label "Quantity"
+      end
+      field "expiration_date" do
+        label "Expiration Date"
+      end
+      field "daysExpiration" do
+        label "Days Until Expiration"
+      end
+    end
```
[the remark this answers is not in the document]
Comment (gopinathmahakud1) [reply cost/azure/reserved_instances/mca_expiration/azure_reserved_instance_expiration.pt]: I have addressed.

File: cost/azure/reserved_instances/mca_recommendations/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+This policy uses [credentials](https://docs.rightscale.com/policies/users/guides/credential_management.html)
+for connecting to the cloud -- in order to apply this policy, you must have a credential registered in the system that is compatible with this policy. If there are no
+credentials listed when you apply the policy, please contact your cloud admin, and ask them to register a credential that is compatible with this policy. The information below should be consulted when creating the credential.
+
+### Credential configuration
+
+For administrators [creating and managing credentials](https://docs.rightscale.com/policies/users/guides/credential_management.html) to use with this policy, the following information is needed: 
+
+Provider tag value to match this policy: `azure`
```
[the remark this answers is not in the document]
Comment (gopinathmahakud1) [reply cost/azure/reserved_instances/mca_recommendations/README.md]: I have addressed.

File: cost/azure/reserved_instances/mca_recommendations/azure_reserved_instance_recommendations.pt [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+# Policy
+###############################################################################
+
+policy "azure_ri_recommendations" do
+  validate_each $ds_ri_normalization do
+    summary_template "{{ len data }} Azure MCA Reserved Instances Purchase Recommendations"
+    escalate $email
+    check lt(to_n(val(item,"netSavings")),$param_savings_threshold)
+    export do
+      resource_level true
```
[the remark this answers is not in the document]
Comment (gopinathmahakud1) [reply cost/azure/reserved_instances/mca_recommendations/azure_reserved_instance_recommendations.pt:165]: I have addressed.

File: cost/azure/reserved_instances/mca_expiration/azure_reserved_instance_expiration.pt [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+# Policy
+###############################################################################
+
+policy "ri_expiration" do
+  validate $ds_filtered_reservations do
+    summary_template "{{ rs_org_name }} (Org ID: {{ rs_org_id }}): {{ len data }} Reserved Instance(s) in Microsoft Azure MCA Expiring"
+    escalate $report_expiring_reserved_instances
+    check eq(size(data),0)
+    export do
+      resource_level true
```
[the remark this answers is not in the document]
Comment (gopinathmahakud1) [reply cost/azure/reserved_instances/mca_expiration/azure_reserved_instance_expiration.pt]: I have addressed.

**Comment by cdwilhelm** [comment]
@gopinathmahakud1   thank you. I'm going to get some feedback from the customer.

**Comment by gopinathmahakud1** [comment]
@cdwilhelm Do you have any customer feedback for this?

**Comment by cdwilhelm** [comment]
> @cdwilhelm Do you have any customer feedback for this?

Very little feedback so far.  There is an issue with one policy which I'm trying to help them on.  I'll get back to you in the next day or two.

**Review by cdwilhelm — APPROVED** [review]

**Pull request merged**